openstack/standalone-cinder/provisioner.go

```diff
@@ -0,0 +1,242 @@
+	// provisioner's PVs.
+	identity string
+}
+
+func newCinderProvisioner(client kubernetes.Interface, id, configFilePath string) (controller.Provisioner, error) {
+	volumeService, err := getVolumeService(configFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get volume service: %v", err)
+	}
+
+	return &cinderProvisioner{
+		volumeService: volumeService,
+		client:        client,
+		identity:      id,
+	}, nil
+}
+
+var _ controller.Provisioner = &cinderProvisioner{}
+
+type provisionCtx struct {
+	p          *cinderProvisioner
+	options    controller.VolumeOptions
+	connection volumeConnection
+}
+
+type deleteCtx struct {
+	p  *cinderProvisioner
+	pv *v1.PersistentVolume
+}
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *cinderProvisioner) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+	if options.PVC.Spec.Selector != nil {
+		return nil, fmt.Errorf("claim Selector is not supported")
+	}
+
+	volumeID, err := createCinderVolume(p, options)
+	if err != nil {
+		glog.Errorf("Failed to create volume")
+		return nil, err
+	}
+
+	err = waitForAvailableCinderVolume(p, volumeID)
+	if err != nil {
+		glog.Errorf("Volume did not become available")
+		return nil, err
+	}
+
+	err = reserveCinderVolume(p, volumeID)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to reserve volume: %v", err)
+		return nil, err
+	}
+
+	connection, err := connectCinderVolume(p, volumeID)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to connect volume: %v", err)
+		return nil, err
+	}
+
+	mapper, err := newVolumeMapperFromConnection(connection)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Unable to create volume mapper: %f", err)
+		return nil, err
+	}
+
+	ctx := provisionCtx{p, options, connection}
+	err = mapper.AuthSetup(ctx)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to prepare volume auth: %v", err)
+		return nil, err
+	}
+
+	pv, err := buildPV(mapper, ctx, volumeID)
+	if err != nil {
+		// TODO: Create placeholder PV?
+		glog.Errorf("Failed to build PV: %v", err)
+		return nil, err
+	}
+
+	return pv, nil
+}
+
+// Delete removes the storage asset that was created by Provision represented
+// by the given PV.
+func (p *cinderProvisioner) Delete(pv *v1.PersistentVolume) error {
+	ann, ok := pv.Annotations[provisionerIDAnn]
+	if !ok {
+		return errors.New("identity annotation not found on PV")
+	}
+	if ann != p.identity {
+		return &controller.IgnoredError{
+			Reason: "identity annotation on PV does not match ours",
+		}
+	}
+	// TODO when beta is removed, have to check kube version and pick v1/beta
+	// accordingly: maybe the controller lib should offer a function for that
+
+	volumeID, ok := pv.Annotations[cinderVolumeID]
+	if !ok {
+		return errors.New("cinder volume id annotation not found on PV")
```

Comment:
add volumeID in the message

Reply:
Changed to `errors.New(cinderVolumeID + " annotation not found on PV")`